### ISSUE TITLE
Worrying iptables errors

### ISSUE THREAD

**Issue by dani-jozsef** [jamesmcm/vopono#66]
(At first i thought that the "use bridge" configuration is being ignored. I realized that while the mullvad ipleak tool doesn't show it, it's connecting to a different server than the exit server. Still, more info in the documentation would be helpful, maybe more choice about the entry server location...)

However, in looking into this topic, I noticed that when pulling up a new namespace, some strange iptables errors happen - maybe it's expecting a different version of iptables? (I'm on Ubuntu 20.04, so it shouldn't be too outdated...  iptables version is `iptables v1.8.4 (legacy)`)

```
$ vopono exec --provider mullvad --server sweden "google-chrome"
 2021-03-07T16:02:30.112Z INFO  vopono::util > Calling sudo for elevated privileges, current user will be used as default user
 2021-03-07T16:02:30.211Z INFO  vopono::util > Chosen config: /home/dani/.config/vopono/mv/wireguard/sweden-se9.conf
 2021-03-07T16:02:30.214Z INFO  vopono::netns > Created new network namespace: vopono_mv_sweden
 2021-03-07T16:02:30.262Z INFO  vopono::netns > IP address of namespace as seen from host: 10.200.1.2
 2021-03-07T16:02:30.262Z INFO  vopono::netns > IP address of host as seen from namespace: 10.200.1.1
Bad argument `addrtype'
Try `iptables -h' or 'iptables --help' for more information.
Bad argument `addrtype'
Try `ip6tables -h' or 'ip6tables --help' for more information.
 2021-03-07T16:02:30.575Z INFO  vopono::exec  > Application google-chrome launched in network namespace vopono_mv_sweden with pid 93266

(google-chrome:93267): dbind-WARNING **: 17:02:30.823: Couldn't connect to accessibility bus: Failed to connect to socket /tmp/dbus-UPvZ3RgVN4: Connection refused
[93304:93304:0307/170230.874610:ERROR:sandbox_linux.cc(374)] InitializeSandbox() called with multiple threads in process gpu-process.
[93556:20:0307/170327.411124:ERROR:stun_port.cc(96)] Binding request timed out from [0:0:0:x:x:x:x:x]:37352 (any)
[93556:20:0307/170825.558446:ERROR:stun_port.cc(96)] Binding request timed out from [0:0:0:x:x:x:x:x]:57085 (any)
[93556:20:0307/171446.541568:ERROR:stun_port.cc(96)] Binding request timed out from [0:0:0:x:x:x:x:x]:48301 (any)
```

**Comment by dani-jozsef** [comment]
These are the offending iptables calls, from vopono -v...

```
 2021-03-07T19:27:52.688Z DEBUG vopono::netns             > ip netns exec vopono_mv_sweden iptables -t nat -A PREROUTING ! -i mv_sweden -d 10.71.107.104/32 addrtype ! --src-type LOCAL -j REJECT
Bad argument `addrtype'
Try `iptables -h' or 'iptables --help' for more information.
 2021-03-07T19:27:52.690Z DEBUG vopono::netns             > ip netns exec vopono_mv_sweden ip6tables -t nat -A PREROUTING ! -i mv_sweden -d fc00:bbbb:bbbb:bb01::8:6b67/128 addrtype ! --src-type LOCAL -j REJECT
Bad argument `addrtype'
Try `ip6tables -h' or 'ip6tables --help' for more information.
```

**Comment by jamesmcm** [comment]
Thanks, this is probably a bug in https://github.com/jamesmcm/vopono/commit/c5f1fa311cd29c1567f5a24a12cc1aaf3acdbaad#diff-2ae045b4ad2239557aacec0570ee0c6f23b9496f756e25b03a904464bf1c3a21

I used nftables, and Wireguard always used nftables before too.

I'll try to check it in detail, but in the meantime I'd recommend trying nftables if you can.

**Comment by dani-jozsef** [comment]
Ubuntu 20.04 is still on the old packet filter backend, so as far as I understand that's not really feasible without upgrading the whole distro.
What's the purpose of these two rules? I might be able to submit a fix, but I'd like to fully understand what's going on. :)

**Comment by jamesmcm** [comment]
I think just removing the `-t nat` might work, I'll try to test it, at least we know the nftables ones work so it's just a conversion.

The aim is to force the traffic over the Wireguard connection, it comes from wg-quick, so basically dropping any connection not on the correct interface if I recall correctly.

Looking at the [patch that swapped wg-quick to nftables](https://lists.zx2c4.com/pipermail/wireguard/2019-December/004727.html), it seems we want is:

```
iptables6 -I PREROUTING ! -i {interface} -d {destination_address} -m addrtype ! --src-type LOCAL -j DROP
```

I think the `-t nat` was included by mistake when copying from other rules (for the whole Command syntax). We might also need to adjust the `-A` for `-I` if the priority is wrong but that's easy to test at least.

I'll try to test it tonight if I can.

But I'd recommend switching to nftables in the future, as with [libvopono](https://github.com/jamesmcm/libvopono), I might eventually have to drop iptables support unless I find an easy way to do it with netlink like nftables. I'll try to keep it though, as the Raspberry Pi is also in the same position at the moment.

**Comment by dani-jozsef** [comment]
Okay I did some testing, but to be honest as I said my iptables is quite rusty... :D

First bad news: doing a naive retooling of Ubuntu 20.04 to nftables results in vopono simply not working anymore. I installed nftables, set all the iptables alternatives to the nftables compatibility layer, rebooted the machine, and while networking was working fine, vopono sandboxed apps had no connectivity for some reason. No errors shown, just no connectivity.

So back to iptables. Tables has a few built-in restrictions.
* There is no PREROUTING chain in the _filter_ table.
* The REJECT target is only allowed in the _filter_ table.
* The DROP target is expressly disallowed in the _nat_ table.

My intuition tells me that this rule should go in the _filter_ table, but then the question is, which chain it should go in - INPUT, OUTPUT or FORWARD...

So... looking at the commands...
```
ip netns exec vopono_mv_sweden iptables -t nat -A PREROUTING ! -i mv_sweden -d 10.71.107.104/32 -m addrtype ! --src-type LOCAL -j REJECT
ip netns exec vopono_mv_sweden ip6tables -t nat -A PREROUTING ! -i mv_sweden -d fc00:bbbb:bbbb:bb01::8:6b67/128 -m addrtype ! --src-type LOCAL -j REJECT
```

What I get out of this is the following: Disallow a packet...
* As soon as a packet enters the packet filter (PREROUTING)
* If it is not assigned to the mv_sweden interface and
* If its destiation is the mv_sweden interface's address and
* If it is not from a local interface address

I'm trying to think of what the attack / bug this is trying to prevent looks like. We're expressly filtering for not local packets, which means this is about external packets coming in through other interfaces? So then it might go in the INPUT chain in the _filter_ table I guess? In that case, the following is accepted by iptables without hiccups:

```
ip netns exec vopono_mv_sweden iptables -A INPUT ! -i mv_sweden -d 10.71.107.104/32 -m addrtype ! --src-type LOCAL -j REJECT
ip netns exec vopono_mv_sweden ip6tables -A INPUT ! -i mv_sweden -d fc00:bbbb:bbbb:bb01::8:6b67/128 -m addrtype ! --src-type LOCAL -j REJECT
```

The question remains, does this actually do what we're expecting of it?

**Comment by dani-jozsef** [comment]
Okay so as far as I could gather, this is the corresponding part from wg-quick:
```
local marker="-m comment --comment \"wg-quick(8) rule for $INTERFACE\"" restore=$'*raw\n'
for i in "${ADDRESSES[@]}"; do
	[[ ( $proto == -4 && $i != *:* ) || ( $proto == -6 && $i == *:* ) ]] || continue
	printf -v restore '%s-I PREROUTING ! -i %s -d %s -m addrtype ! --src-type LOCAL -j DROP %s\n' "$restore" "$INTERFACE" "${i%/*}" "$marker"
done
```
It seems to be inserting the rule in the _PREROUTING_ chain of the _raw_ table. So I'm pretty certain that it is supposed to be
```
ip netns exec vopono_mv_sweden iptables -t raw -I PREROUTING ! -i mv_sweden -d 10.71.107.104/32 -m addrtype ! --src-type LOCAL -j DROP
ip netns exec vopono_mv_sweden ip6tables -t raw -I PREROUTING ! -i mv_sweden -d fc00:bbbb:bbbb:bb01::8:6b67/128 -m addrtype ! --src-type LOCAL -j DROP
```

**Comment by jamesmcm** [comment]
Thanks for the PR, this was merged in [version 0.6.8](https://github.com/jamesmcm/vopono/releases/tag/0.6.8).

**Comment by dani-jozsef** [comment]
Nice, thanks!
